File: shopping_android_app/app/src/main/java/com/example/shopping_android_app/ui/home/HomeFragment.java
package com.example.shopping_android_app.ui.home;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.bumptech.glide.Glide;
import com.example.shopping_android_app.MainActivity;
import com.example.shopping_android_app.R;
import com.example.shopping_android_app.adapter.BrandListAdapter;
import com.example.shopping_android_app.adapter.CategoryListAdapter;
import com.example.shopping_android_app.adapter.HotGoodsAdapter;
import com.example.shopping_android_app.adapter.NewGoodsListAdapter;
import com.example.shopping_android_app.adapter.TopicListAdapter;
import com.example.shopping_android_app.base.BaseAdapter;
import com.example.shopping_android_app.base.BaseFragment;
import com.example.shopping_android_app.interfaces.home.IHome;
import com.example.shopping_android_app.model.home.CategoryBean;
import com.example.shopping_android_app.model.home.CategoryListBean;
import com.example.shopping_android_app.model.home.HomeBean;
import com.example.shopping_android_app.presenter.home.HomePresenter;
import com.example.shopping_android_app.ui.channel.CategoryActivity;
import com.example.shopping_android_app.ui.dashboard.BrandActivity;
import com.example.shopping_android_app.ui.dashboard.DashboardFragment;
import com.example.shopping_android_app.ui.details.CarActivity;
import com.youth.banner.Banner;
import com.youth.banner.loader.ImageLoader;

import java.util.List;

import butterknife.BindView;
import butterknife.OnClick;

public class HomeFragment extends BaseFragment<HomePresenter> implements IHome.View {

    @BindView(R.id.action_up_btn)
    Button actionUpBtn;
    @BindView(R.id.linesr)
    LinearLayout linesr;
    @BindView(R.id.liness)
    LinearLayout liness;
    @BindView(R.id.banner)
    Banner banner;
    @BindView(R.id.recy_brand)
    RecyclerView recyBrand;
    @BindView(R.id.recy_newgood)
    RecyclerView recyNewgood;
    @BindView(R.id.recy_pop)
    RecyclerView recyPop;
    @BindView(R.id.recy_tit)
    RecyclerView recyTit;
    @BindView(R.id.txt_brand_title)
    TextView txtBrandTitle;
    @BindView(R.id.txt_newgood_title)
    TextView txtNewgoodTitle;
    @BindView(R.id.txt_pop_title)
    TextView txtPopTitle;
    @BindView(R.id.txt_tit_title)
    TextView txtTitTitle;

    @Override
    protected HomePresenter createPersenter() {
        return new HomePresenter();
    }

    @Override
    protected void initData() {
        presenter.getHome();
    }

    @Override
    protected void initView() {
        txtTitTitle.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(getActivity(), MainActivity.class);
                intent.putExtra("id",1);
                startActivity(intent);
            }
        });
    }

    @Override
    protected int getLayout() {
        return R.layout.fragment_home;
    }

    @Override
    public void showLoading(int visible) {

    }

    @Override
    public void showToast(String tips) {

    }

    @Override
    public void getHomeReturn(HomeBean result) {
        if (result != null) {
            HomeBean.DataBean data = result.getData();
            initBanner(data.getBanner());
            initChannel(data.getChannel());
            initBrand(data.getBrandList());
            initNewGoodsList(data.getNewGoodsList());
            initHotGoodsList(data.getHotGoodsList());
            initTopicList(data.getTopicList());
            initCategoryList(data.getCategoryList());
        }
    }

    @Override
    public void getCategory(CategoryBean categoryBean) {

    }

    @Override
    public void getCategoryList(CategoryListBean categoryListBean) {

    }

    private void initChannel(List<HomeBean.DataBean.ChannelBean> channel) {
        for (int i = 0; i < channel.size(); i++) {
            View view = LayoutInflater.from(getActivity()).inflate(R.layout.home_item_ssss, null);
            TextView text = view.findViewById(R.id.txt_home_title);
            text.setText(channel.get(i).getName());
            ImageView image = view.findViewById(R.id.image);

            Glide.with(getActivity()).load(channel.get(i).getIcon_url()).into(image);
            view.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT, 1.0f));
            liness.addView(view);
            view.setTag(channel.get(i));
            view.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    String url = ((HomeBean.DataBean.ChannelBean) v.getTag()).getUrl();
                    String name = ((HomeBean.DataBean.ChannelBean) v.getTag()).getName();
                    int i1 = url.indexOf("=");
                    String substring = url.substring(i1 + 1);
                    Intent intent = new Intent(getActivity(), CategoryActivity.class);
                    intent.putExtra("name", name);
                    intent.putExtra("url", substring);
                    startActivity(intent);
                }
            });
        }
    }

    private void initCategoryList(List<HomeBean.DataBean.CategoryListBean> goodsList) {
        for (int i = 0; i < goodsList.size(); i++) {
            View view = LayoutInflater.from(getActivity()).inflate(R.layout.home_item_sss, null);
            TextView text = view.findViewById(R.id.txt_home_title);
            text.setText(goodsList.get(i).getName());
            RecyclerView recy_home = view.findViewById(R.id.recy_home);
            List<HomeBean.DataBean.CategoryListBean.GoodsListBean> goodsList1 = goodsList.get(i).getGoodsList();
            recy_home.setLayoutManager(new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL));
            CategoryListAdapter categoryListAdapter = new CategoryListAdapter(getActivity(), goodsList1);
            recy_home.setAdapter(categoryListAdapter);
            view.setTag(goodsList1);
            linesr.addView(view);
            categoryListAdapter.addListClick(new BaseAdapter.IListClick() {
                @Override
                public void itemClick(int pos) {
                    List<HomeBean.DataBean.CategoryListBean.GoodsListBean> goodsListBeans = (List<HomeBean.DataBean.CategoryListBean.GoodsListBean>) view.getTag();
                    HomeBean.DataBean.CategoryListBean.GoodsListBean goodsListBean = goodsListBeans.get(pos);
                    Intent intent = new Intent(getActivity(), CarActivity.class);
                    intent.putExtra("id", goodsListBean.getId());
                    startActivity(intent);
                }
            });
        }

    }


    private void initTopicList(List<HomeBean.DataBean.TopicListBean> topicList) {
        recyTit.setLayoutManager(new StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.HORIZONTAL));

        TopicListAdapter topicListAdapter = new TopicListAdapter(getActivity(), topicList);

        recyTit.setAdapter(topicListAdapter);
    }

    private void initHotGoodsList(List<HomeBean.DataBean.HotGoodsListBean> hotGoodsList) {
        recyPop.setLayoutManager(new LinearLayoutManager(getActivity()));

        recyPop.addItemDecoration(new DividerItemDecoration(getActivity(), DividerItemDecoration.VERTICAL));

        HotGoodsAdapter hotGoodsAdapter = new HotGoodsAdapter(getActivity(), hotGoodsList);

        recyPop.setAdapter(hotGoodsAdapter);

    }

    private void initBrand(List<HomeBean.DataBean.BrandListBean> brandList) {
        recyBrand.setLayoutManager(new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL));

        BrandListAdapter brandListAdapter = new BrandListAdapter(getActivity(), brandList);

        recyBrand.setAdapter(brandListAdapter);
    }

    private void initNewGoodsList(List<HomeBean.DataBean.NewGoodsListBean> newGoodsList) {

        recyNewgood.setLayoutManager(new StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL));

        NewGoodsListAdapter newGoodsListAdapter = new NewGoodsListAdapter(getActivity(), newGoodsList);

        recyNewgood.setAdapter(newGoodsListAdapter);
    }

    private void initBanner(List<HomeBean.DataBean.BannerBean> list) {
        banner.setImages(list).setImageLoader(new ImageLoader() {
            @Override
            public void displayImage(Context context, Object path, ImageView imageView) {
                HomeBean.DataBean.BannerBean bannerBean = (HomeBean.DataBean.BannerBean) path;
                Glide.with(context).load(bannerBean.getImage_url()).into(imageView);
            }
        }).start();
    }


    @OnClick({R.id.txt_brand_title, R.id.txt_newgood_title})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.txt_brand_title:
                startActivity(new Intent(getActivity(), BrandActivity.class));
                break;
            case R.id.txt_newgood_title:
                startActivity(new Intent(getActivity(), HotGoodActivity.class));
                break;
        }
    }
}

File: kotlin/app/src/main/java/com/example/myapplication/home/HomeFragment.kt
package com.example.myapplication.home


import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Color
import android.util.SparseArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.databinding.library.baseAdapters.BR
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.bumptech.glide.Glide
import com.example.myapplication.R
import com.example.myapplication.adapter.home.*
import com.example.myapplication.base.IItemClick
import com.example.myapplication.data.*
import com.example.myapplication.databinding.FragmentHomeBinding
import com.example.myapplication.tongpao.HotActivity
import com.example.myapplication.viewmodel.BindHomeViewModel
import com.shop.base.BaseFragment
import kotlinx.android.synthetic.main.fragment_home.*
import kotlinx.android.synthetic.main.fragment_home.view.*
import kotlinx.android.synthetic.main.home_item_category.view.*
import kotlinx.android.synthetic.main.home_item_channel.view.*
import kotlinx.android.synthetic.main.home_item_channel.view.txt_home_title


class HomeFragment : BaseFragment<BindHomeViewModel, FragmentHomeBinding>(
    R.layout.fragment_home,
    BindHomeViewModel::class.java
),View.OnClickListener {


    //点击事件
    private fun initClick() {
        //品牌制作商直供  点击事件
//        mDataBinding!!.txtBrandTitle.setOnClickListener(this)
//        mDataBinding!!.txtNewgoodTitle.setOnClickListener(this)
//        mDataBinding!!.txtTitTitle.setOnClickListener(this)
    }

    //点击监听
    override fun onClick(v: View?) {
        when(v?.id){
            R.id.txt_brand_title -> startActivity(Intent(context, BrandActivity::class.java))
            R.id.txt_newgood_title -> startActivity(Intent(context, NewGoodsActivity::class.java))
            R.id.txt_tit_title -> startActivity(Intent(context, HotActivity::class.java))

        }
    }

    override fun initView() {
        mViewModel!!.getHome()
        //banner图
        showBanner()
        //商品类型
        showChannel()
        //品牌制造商
        showBrand()
        //新品首发
        showNewGoodsList()
        //人气推荐
        showHotGoodsList()
        //专题
        showTopicList()
        //居家
        showCategoryList()
        mDataBinding!!.setVariable(BR.homeClick, TitleClick())
    }

    //居家
    private fun showCategoryList() {
        mViewModel!!.category.observe(this, {
            for (item in it) {
                var view = LayoutInflater.from(context).inflate(R.layout.home_item_category, null)
                var binging: ViewDataBinding? = DataBindingUtil.bind(view)
                binging?.setVariable(BR.vmCateGory, item)
                view!!.recy_category.layoutManager = StaggeredGridLayoutManager(
                    2,
                    StaggeredGridLayoutManager.VERTICAL
                )
                var layouts = SparseArray<Int>()
                layouts.put(R.layout.home_item_categorygoods,BR.vmGoods)

                var categoryListAdapter = CategoryListAdapter(context, item.goodsList,layouts,ItemClick())
                view!!.recy_category.adapter = categoryListAdapter
                initCateGory(item.goodsList, categoryListAdapter)
                view!!.setTag(item.goodsList)
                mDataBinding!!.linesr.addView(view)
            }
        })
    }

    private fun initCateGory(goodsList: List<Goods>, categoryListAdapter: CategoryListAdapter) {
        categoryListAdapter.refreshData(goodsList)
    }


    //专题
    private fun showTopicList() {
        var layout = StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.HORIZONTAL)
        mDataBinding!!.recyTopic.layoutManager = layout
        var topicListAdapter = TopicListAdapter(context)
        mDataBinding!!.recyTopic.adapter=topicListAdapter
        initTopic(topicListAdapter)
    }

    private fun initTopic(topicListAdapter: TopicListAdapter) {
        mViewModel!!.topic.observe(this, Observer {
            topicListAdapter.refreshData(it)
        })
    }

    //人气推荐
    private fun showHotGoodsList() {
        var layout = StaggeredGridLayoutManager(1, StaggeredGridLayoutManager.VERTICAL)
        mDataBinding!!.recyHotgoods.layoutManager = layout
        var hotGoodsAdapter = HotGoodsAdapter(context)
        mDataBinding!!.recyHotgoods.adapter=hotGoodsAdapter
        initHotGoods(hotGoodsAdapter)
    }

    private fun initHotGoods(hotGoodsAdapter: HotGoodsAdapter) {
        mViewModel!!.hot.observe(this, Observer {
            hotGoodsAdapter.refreshData(it)
        })
    }

    //新品首发
    private fun showNewGoodsList() {
        var layout = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        mDataBinding!!.recyNewgood.layoutManager = layout
        var newGoodsAdapter = NewGoodsAdapter(context)
        mDataBinding!!.recyNewgood.adapter=newGoodsAdapter
        initNewGoods(newGoodsAdapter)
    }

    private fun initNewGoods(newGoodsAdapter: NewGoodsAdapter) {
        mViewModel!!.newGoods.observe(this, Observer {
            newGoodsAdapter.refreshData(it)
        })
    }

    //品牌制造商
    private fun showBrand() {

        var layout = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)
        mDataBinding!!.recyBrand.layoutManager =layout
        var brandAdapter = BrandAdapter(context)
        mDataBinding!!.recyBrand.adapter = brandAdapter
        initVM(brandAdapter)
    }

    //刷新适配器
    private fun initVM(baseAdapter: BrandAdapter) {
        mViewModel!!.brand.observe(this, Observer {
            baseAdapter.refreshData(it)
        })
    }


    //商品类型
    private fun showChannel() {
        mViewModel!!.channel.observe(this, {
            for (i in it.indices) {
                //获取布局
                val view: View = LayoutInflater.from(activity).inflate(
                    R.layout.home_item_channel,
                    null
                )
                Glide.with(this).load(it.get(i).icon_url).into(view.image)
                view.txt_home_title.setText(it.get(i).name)

                view.layoutParams = LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    1.0f
                )

                mDataBinding!!.linerChannel.addView(view)

                view.setTag(it.get(i))

                //条目点击事件
                view.setOnClickListener { v ->
                    //强转 as
                    val views = view.tag as Channel
                    val url: String = views.url
                    val name: String = views.name
                    val i1 = url.indexOf("=")
                    val substring = url.substring(i1 + 1)
                    Toast.makeText(context, substring, Toast.LENGTH_LONG).show()
                    var intent = Intent(context,CateGoryActivity::class.java)
                    intent.putExtra("id",substring)
                    intent.putExtra("name",name)
                    startActivity(intent)
                }
            }
        })



    }

    //banner图
    private fun showBanner() {
        mViewModel!!.bannner.observe(this, Observer {
            mDataBinding!!.bannerHome.setImages(it)
            mDataBinding!!.bannerHome.setImageLoader(MyBannerAdapter())
            mDataBinding!!.bannerHome.start()
        })

    }

    override fun initVM() {

    }

    override fun initData() {

    }

    override fun initVariable() {
    }

    inner class ItemClick : IItemClick<Goods>{
        override fun itemClick(data: Goods) {
            var intent = Intent(context,CarActivity::class.java)
            intent.putExtra("id",data.id.toString())
            startActivity(intent)
        }

    }

    inner class TitleClick{
        //品牌直供
        fun clickBrand(){
            startActivity(Intent(context, BrandActivity::class.java))
        }
        //新品发布
        fun clickNewGood(){
            startActivity(Intent(context, NewGoodsActivity::class.java))
        }
        //人气推荐
        fun clickHotGood(){

        }
        //专题
        fun clickTopic(){
            startActivity(Intent(context, HotActivity::class.java))
        }
    }
}